addons/print/models/print_printer.py

```diff
@@ -195,70 +251,66 @@ def spool_report(self, docids, report_name, data=None, title=None, copies=1):
                 reports = self.env.ref(name, raise_if_not_found=False)
             if not reports:
                 raise UserError(_("Undefined report %s") % name)
-
-        # Identify required report types
-        report_types = set(reports.mapped("report_type"))
-
-        # If there is only 1 report type to print out then raise an error if printer not found,
-        # otherwise continue checking for other report types
-        printer_raise_if_not_found = len(report_types) == 1
-
-        # Identify printer to use for each report type, only include report type in dictionary
-        # if a printer is identified
+        return reports
+
+    def _assign_printers_to_report(self, required_report_types):
+        """
+        Purpose: Assign each printer in self (or system default if self is empty) to a given
+        report type.
+        Return: dict{ ir.action.report() : print.printer() }
+        """
+        printer_raise_if_not_found = len(required_report_types) == 1
         printers_by_report_type = {
             rt: p
-            for rt in report_types
-            for p in (self.printers(raise_if_not_found=printer_raise_if_not_found, report_type=rt))
+            for rt in required_report_types
+            for p in (self.printers(rt, printer_raise_if_not_found))
             if p
         }
-
+        return printers_by_report_type
+
+    def _check_every_report_type_has_printer(self, printers_by_report_type, required_report_types):
+        """
+        Purpose: Check which report types are missing, if any are missing raise an error. Must raise an
+        error as can't print the given report type.
+        Return: Boolean
+        """
         if printers_by_report_type:
-            required_types = set(printers_by_report_type.keys())
-            missing = required_types - report_types
+            report_types_found = set(printers_by_report_type.keys())
+            missing = required_report_types - report_types_found
         else:
-            missing = report_types
+            missing = required_report_types
 
         if missing:
-            # Get labels of missing report types from selection list
             missing_labels = [self.get_report_type_label(rt) for rt in missing]
             raise UserError(_("Missing reports of types: %s") % ", ".join(missing_labels))
 
-        # CPCL reports require number of copies passed into the template via data
```

Comment:
Might be nice to keep this comment where you placed the code below so that we don't forget why we do this?

Reply:
This one I did keep, it should be on line 214